port/unix/omrsysinfo.c

```diff
@@ -7557,3 +7559,81 @@ omrsysinfo_get_number_context_switches(struct OMRPortLibrary *portLibrary, uint6
 	return OMRPORT_ERROR_SYSINFO_NOT_SUPPORTED;
 #endif /* defined(LINUX) */
 }
+
+/**
+ * Get the process ID and commandline for each process.
+ * @param[in] portLibrary The port library.
+ * @param[in] callback The function to be invoked for each process with the process ID and command info.
+ * @param[in] userData Data passed to the callback.
+ * @return 0 on success, or the first non-zero value returned by the callback.
+ */
+uintptr_t
+omrsysinfo_get_processes(struct OMRPortLibrary *portLibrary, OMRProcessInfoCallback callback, void *userData)
+{
+#if defined(LINUX)
+	DIR *dir = NULL;
+	struct dirent *entry = NULL;
+	char cmdline_path[PATH_MAX];
+	char comm_path[PATH_MAX];
+	char command[PATH_MAX];
+	uintptr_t file = 0;
+	uintptr_t callback_result = 0;
+	size_t bytes_read = 0;
+	dir = opendir("/proc");
+	if (NULL == dir) {
+		int32_t rc = findError(errno);
+		portLibrary->error_set_last_error(portLibrary, errno, rc);
+		Trc_PRT_sysinfo_get_open_file_count_failedOpeningProcFS(rc);
+		return (uintptr_t)-1;
+	}
+	entry = readdir(dir);
+	while (NULL != entry) {
+		/* Skip non-numeric entries which are not processes */
+		if (0 == isdigit(entry->d_name[0])) {
+			entry = readdir(dir);
+			continue;
+		}
+		memset(command, 0, PATH_MAX);
+		portLibrary->str_printf(portLibrary, cmdline_path, sizeof(cmdline_path), "/proc/%s/cmdline", entry->d_name);
+		portLibrary->str_printf(portLibrary, comm_path, sizeof(comm_path), "/proc/%s/comm", entry->d_name);
+		bytes_read = 0;
+		/* Open the cmdline file and read cmdline */
+		file = portLibrary->file_open(portLibrary, cmdline_path, EsOpenRead, 0);
+		if (0 != file) {
+			bytes_read = portLibrary->file_read(portLibrary, file, command, PATH_MAX - 1);
+			portLibrary->file_close(portLibrary, file);
+		}
+		/* If cmdline is empty, try reading from comm */
+		if (0 == bytes_read) {
+			file = portLibrary->file_open(portLibrary, comm_path, EsOpenRead, 0);
+			if (0 != file) {
+				bytes_read = portLibrary->file_read(portLibrary, file, command, PATH_MAX - 1);
+				portLibrary->file_close(portLibrary, file);
+			}
+		}
+		/* If both cmdline and comm are empty, skip the process */
+		if (0 == bytes_read) {
+			entry = readdir(dir);
+			continue;
+		}
+		/* Replacing null terminators with spaces in the command */
+		for (size_t i = 0; i < bytes_read; i++) {
+			if ('\0' == command[i]) {
+				command[i] = ' ';
+			}
+		}
+		/* Call the callback function with the process ID and command */
+		callback_result = callback((uintptr_t)strtoull(entry->d_name, NULL, 10), command, userData);
+		if (0 != callback_result) {
+			closedir(dir);
+			return callback_result;
+		}
+		entry = readdir(dir);
+	}
+	closedir(dir);
+	return 0;
```

Comment:
This can be simplified so only one call to `closedir()` is required:
```c
		if (0 != callback_result) {
			break;
		}
	}
	closedir(dir);
	return callback_result;
```